File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer token;

    static int N, K;
    static boolean[] visited;

    public static void main(String[] args) throws IOException {
        token = new StringTokenizer(input.readLine());
        N = Integer.parseInt(token.nextToken());
        K = Integer.parseInt(token.nextToken());

        visited = new boolean[100001];
        Deque<int[]> deq = new ArrayDeque<>();
        deq.addLast(new int[]{N, 0});
        visited[N] = true;

        while (!deq.isEmpty()) {
            int[] cur = deq.pollFirst();
            int x = cur[0];
            int cnt = cur[1];

            if (x == K) {
                System.out.println(cnt);
                break;
            }

            if (isValid(x*2) && !visited[x*2]) {
                deq.addFirst(new int[]{x*2, cnt});
                visited[x*2] = true;
            }

            if (isValid(x-1) && !visited[x-1]) {
                deq.addLast(new int[]{x-1, cnt+1});
                visited[x-1] = true;
            }

            if (isValid(x+1) && !visited[x+1]) {
                deq.addLast(new int[]{x+1, cnt+1});
                visited[x+1] = true;
            }
        }
    }

    public static boolean isValid(int x) {
        return x >= 0 && x <= 100000;
    }
}

File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.py
'''
아이디어:
0초(순간이동)부터 모두 진행 후, 걷기
-> 순간이동은 deq.appendleft로 먼저 처리
'''
import sys
from collections import deque

def is_valid(x):
    return 0<=x<MAX_LEN

delta_x = [-1,1]

MAX_LEN = 100_001
N,K = map(int, sys.stdin.readline().split())
visited = [False]*MAX_LEN

deq = deque()
deq.append((N,0)) # 시작 위치, 시간
while deq:
    idx, time = deq.popleft()
    if idx == K: # 찾음
        print(time)
        break

    if is_valid(idx*2) and not visited[idx*2]: # 순간이동
        deq.appendleft((idx*2,time)) # 왼쪽에 넣어 먼저 처리
        visited[idx*2] = True

    for k in range(2):
        dx = idx + delta_x[k]
        if is_valid(dx) and not visited[dx]:
            deq.append((dx,time+1))
            visited[dx] = True